cmd/benchrunner.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/elastic/elastic-package/internal/benchrunner/reporters/formats"
+	"github.com/elastic/elastic-package/internal/benchrunner/reporters/outputs"
+	_ "github.com/elastic/elastic-package/internal/benchrunner/runners" // register all benchmark runners
+	"github.com/elastic/elastic-package/internal/cobraext"
+	"github.com/elastic/elastic-package/internal/common"
+	"github.com/elastic/elastic-package/internal/elasticsearch"
+	"github.com/elastic/elastic-package/internal/packages"
+	"github.com/elastic/elastic-package/internal/signal"
+)
+
+const benchLongDescription = `Use this command to run benchmarks on a package. Currently, the following types of benchmarks are available:
+
+#### Pipeline Benchmarks
+These benchmarks allow you to benchmark any Ingest Node Pipelines defined by your packages.
+
+For details on how to configure pipeline benchmarks for a package, review the [HOWTO guide](https://github.com/elastic/elastic-package/blob/main/docs/howto/pipeline_benchmarking.md).`
+
+func setupBenchmarkCommand() *cobraext.Command {
+	var benchTypeCmdActions []cobraext.CommandAction
+
+	cmd := &cobra.Command{
+		Use:   "benchmark",
+		Short: "Run benchmarks for the package",
+		Long:  benchLongDescription,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cmd.Println("Run benchmarks for the package")
+
+			if len(args) > 0 {
+				return fmt.Errorf("unsupported benchmark type: %s", args[0])
+			}
+
+			return cobraext.ComposeCommandActions(cmd, args, benchTypeCmdActions...)
+		}}
+
+	cmd.PersistentFlags().BoolP(cobraext.FailOnMissingFlagName, "m", false, cobraext.FailOnMissingFlagDescription)
+	cmd.PersistentFlags().StringP(cobraext.ReportFormatFlagName, "", string(formats.ReportFormatHuman), cobraext.ReportFormatFlagDescription)
+	cmd.PersistentFlags().StringP(cobraext.ReportOutputFlagName, "", string(outputs.ReportOutputSTDOUT), cobraext.ReportOutputFlagDescription)
+
+	for benchType, runner := range benchrunner.BenchRunners() {
+		action := benchTypeCommandActionFactory(runner)
+		benchTypeCmdActions = append(benchTypeCmdActions, action)
+
+		benchTypeCmd := &cobra.Command{
+			Use:   string(benchType),
+			Short: fmt.Sprintf("Run %s benchmarks", runner.String()),
+			Long:  fmt.Sprintf("Run %s benchmarks for the package.", runner.String()),
+			RunE:  action,
+		}
+
+		benchTypeCmd.Flags().StringSliceP(cobraext.DataStreamsFlagName, "d", nil, cobraext.DataStreamsFlagDescription)
+
+		cmd.AddCommand(benchTypeCmd)
+	}
+
+	return cobraext.NewCommand(cmd, cobraext.ContextPackage)
+}
+
+func benchTypeCommandActionFactory(runner benchrunner.BenchRunner) cobraext.CommandAction {
+	benchType := runner.Type()
+	return func(cmd *cobra.Command, args []string) error {
+		cmd.Printf("Run %s benchmarks for the package\n", benchType)
+
+		failOnMissing, err := cmd.Flags().GetBool(cobraext.FailOnMissingFlagName)
+		if err != nil {
+			return cobraext.FlagParsingError(err, cobraext.FailOnMissingFlagName)
+		}
```

Comment:
I guess it depends. I was wondering if would be practical to add a config option to use the pipeline tests events as samples, to avoid repetition and basically to be able to leverage benchmarks for any package with pipeline tests without changes. If added, would fallback to lookup pipeline test samples, and I guess than in a scenario like this failing could be useful? WDYT?

Reply:
> If added, would fallback to lookup pipeline test samples, and I guess than in a scenario like this failing could be useful

Yes, in this case, it makes sense.

Based on your experience, how many cases would be covered by borrowing pipeline test events? Is it the majority or just w few samples? If that feature doesn't look to be popular, I would rather drop it.

Reply:
I think would be nice to aim to have this in some capacity as part of the CI, so I think all packages would end up using it in one way or another. cc @leehinman